rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+
+Apache Beam mostly follows semantic versioning: breakages are rare and we have
+only seen one occurrence related to type hints once in the entirety of 2020.
+Therefore we propose the following for all TFX 1.x versions:
+
+1.  Each version of TFX is guaranteed to be compatible with latest version of
+    `apache-beam` at release time;
+2.  In addition, `tfx`'s `install_requires` will allow future minor versions of
+    `apache-beam` in the same major version, i.e, `apache-beam>=2.27,<3`.
+3.  When a breaking change to above from `apache-beam` is detected, TFX team
+    will create patch versions to either (a) cherry-pick a fix, or (b) lower the
+    upper bound;
+    1.  For example, say `tfx` 1.1.0 is released with `apache-beam>=2.26,<3`,
+        and `tfx` 1.2.0 is released with `apache-beam>=2.27,<3` (version numbers
+        are examples rather than real). If `apache-beam` 2.30 introduces a
+        breaking change deemed impossible to fix, `tfx` 1.1.1 and 1.2.1 will be
+        released with `apache-beam>=2.27,<2.30`. Please also note that we commit
+        to test tfx with `apache-beam` RC so such breakages should be
+        exceedingly rare in the future.
+
+#### PyArrow {#pyarrow}
+
+TFX’s dependency on PyArrow comes from
+[TFXIO](https://github.com/tensorflow/tfx-bsl/blob/19df0b53e004404ae1971d4140a3242f3490f2d4/tfx_bsl/tfxio/tfxio.py#L35).
+Specific public APIs exposed through are:
+
+*   pyarrow.RecordBatch (thus pyarrow.Array, and more)
+*   pyarrow.Schema
+    *   The stability of contents of the schema (i.e. how data is represented in
+        RecordBatches) is guaranteed by TFX-BSL, please see the “TFX-BSL”
+        section.
+
+For the entire TFX 1.x, at least PyArrow 1.x and 2.x will be supported.
+
+### TFX & ml-pipelines-sdk {#tfx-&-ml-pipelines-sdk}
+
+This section discusses all parts of the `tfx` package, including the minimal SDK
+package _ml-pipelines-sdk_, in detail.
+
+#### Artifact Payload {#artifact-payload}
+
+Almost all supported Artifact types have some payload on-disk storage. The
+compatibility guarantee is the following:
+
+*   Guaranteed Backward compatibility: within same major version, new release of
+    TFX components shall take Artifacts produced by an older version of TFX;
+    *   We shall record versions of all Artifacts: unless otherwise documented,
+        the version of an Artifact equals to the _&lt;major>.&lt;minor>_ version
+        of its producer component. This version will be further used as
+        consumption hint for these Artifacts, to correctly handle version
+        upgrades.
+*   **No Forward compatibility guarantee on payload**: while the system will not
+    reject passing Artifacts produced from newer versions of TFX to an old
+    version of a component, we do not guarantee this will work. A warning will
+    be generated in this case.
+*   Resolvers and importers would only resolve Artifacts produced with same or
+    older versions of the library, by default.
```

Comment:
What kind of warning you are looking for? If 1) all candidate artifacts are from newer version of the library and 2) the semantic  of the resolver and importer used here cannot tolerate empty result, we would see a hard error.

It seems like the only warning we could add is something like following in resolver or importer execution:

```python
if skipped_artifacts:
  logging.warn('Skipped %d artifacts because they are produced by future version of the component', len(skipped_artifacts)')
```

Reply:
Yeah, that would be enough 👍

Reply:
Added to the doc.